nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/SameDiffLoss.java

```diff
@@ -0,0 +1,153 @@
+     * @param preOutput    Output of the model (neural network)
+     * @param activationFn Activation function that should be applied to preOutput
+     * @param mask         Mask array; may be null
+     * @param average      Whether the score should be averaged (divided by number of rows in labels/preOutput) or not   @return Loss function value
+     */
+    public double computeScore(INDArray labels, INDArray preOutput, IActivation activationFn, INDArray mask, boolean average) {
+
+        // The score overall consists of the
+        // sum of the negative log likelihoods for each
+        // of the individual labels.
+
+
+        INDArray scoreArr = computeScoreArray(labels, preOutput, activationFn, mask);
+
+        double score = scoreArr.sumNumber().doubleValue();
+        if (average) {
+            score /= scoreArr.size(0);
+        }
+        return score;
+    }
+
+
+    /**
+     * Compute the score (loss function value) for each example individually.
+     * For input [numExamples,nOut] returns scores as a column vector: [numExamples,1]
+     *
+     * @param labels       Labels/expected output
+     * @param preOutput    Output of the model (neural network)
+     * @param activationFn Activation function that should be applied to preOutput
+     * @param mask         @return Loss function value for each example; column vector
+     */
+    public INDArray computeScoreArray(INDArray labels, INDArray preOutput, IActivation activationFn, INDArray mask) {
+
+        Preconditions.checkArgument((labels.size(1) != preOutput.size(1)), "Labels array numColumns (size(1) = %s) does not match output layer number of outputs (nOut = %s)", labels.size(1), preOutput.size(1));
+        INDArray scoreArr;
+        INDArray output = activationFn.getActivation(preOutput.dup(), true);
+        sd.output(Collections.singletonMap("output", output));
+        scoreArr = output.rsubi(labels).divi(labels);
+        scoreArr.muli(100.0 / labels.size(1));
+
+
+        if (mask != null) {
+            LossUtil.applyMask(scoreArr, mask);
+        }
+        return scoreArr.sum(1);
+
+
+    }
+
+
+    /**
+     * Compute the gradient of the loss function with respect to the inputs: dL/dOutput
+     *
+     * @param labels       Label/expected output
+     * @param preOutput    Output of the model (neural network), before the activation function is applied
+     * @param activationFn Activation function that should be applied to preOutput
+     * @param mask         Mask array; may be null
+     * @return Gradient dL/dPreOut
+     */
+    public INDArray computeGradient(INDArray labels, INDArray preOutput, IActivation activationFn, INDArray mask) {
+
+
+        Map<String, INDArray> grads = sd.calculateGradients(Collections.singletonMap("out", activationFn.getActivation(preOutput.dup(), true)));
+
+        if (mask != null) {
+            LossUtil.applyMask(sd.getVariable("out").eval(), mask);
+        }
+        return sd.getVariable("out").eval();
```

Comment:
You should be applying the mask to - and the returning - the "layerInput" value from `Map<String, INDArray> grads`